src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -279,8 +280,20 @@ public Task OnDataAsync(Http2Frame dataFrame)
             {
                 RequestBodyStarted = true;
 
+                if (InputRemaining.HasValue)
+                {
+                    // https://tools.ietf.org/html/rfc7540#section-8.1.2.6
+                    if (payload.Count > InputRemaining.Value)
+                    {
+                        throw new Http2StreamErrorException(StreamId, CoreStrings.Http2StreamErrorMoreDataThanLength, Http2ErrorCode.PROTOCOL_ERROR);
```

Comment:
Does this abort the connection or just the stream? If the latter, I'm pretty sure the payload still needs to count towards the connection-level flow control window. Calling  `_inputFlowControl.Abort()` _then_ `_inputFlowControl.Advance(payload.Count)` should do the trick.

Reply:
It's only for the stream. I'll take a look at flow connection flow control accounting.